magento/pwa-studio
# [bugfix]: Fixes hot reloading

## Description

Before I begin let's reproduce this so you know what I'm solving here.

1) On `develop`, open your local dev server with `yarn watch:all` and navigate to the `/create-account` page.
2) Now open `createAccountPage.js` and add `import Button from '../Button';` to the imports and `<Button>Foo</Button>` just beneath the container `div`. 
3) When you save this file and the app should start reloading and you should see an error in console (`staticContext` warning is resolved in #1966) _AND_ the page should look completely broken:

[![Image from Gyazo](https://i.gyazo.com/aa216c5e104f76bdb0aacb037424ec97.png)](https://gyazo.com/aa216c5e104f76bdb0aacb037424ec97)

4) Now, check out this PR (reset your local changes too) and re-run the steps. The page should render and you should see the button appear without a reload!

**So what went wrong?**

This was an elusive [hot reloading](https://webpack.js.org/api/hot-module-replacement/) bug that was difficult to reproduce. I was able to finally repro by making changes to direct child of `Route` that was being lazy loaded. I assumed it was something to do with hot reloading as you could restart the watcher and the page would render fine (until you made some changes). The odd thing was that you could get the `Button` component "shell" to display but the child components and any other components on that page such as the `CreateAccountForm` would just not render.

After poking around I found [this comment](https://github.com/gaearon/react-hot-loader/issues/1139#issuecomment-449897132) on an issue on the "React hot loader" plugin. The comment suggests that React does some "double buffering" where it displays/updates an _alternate_ before displaying the final state. This sounded a lot like what was happening (although I can't say 100% without diving into React internals), particularly because of the "partial" render state that was returned after making changes.

I also noticed that our hot reloading of React components was not actaully hot reloading. In fact it seemed to be forcing a refresh after each change instead of swapping out the modified component.

So, to address the above issues I did the following:

1) I added [react-hot-loader](https://github.com/gaearon/react-hot-loader) to the babel plugins for `development`. Now, changes to lazy load components should work!
2) I added some additional code necessary to actually hot reload our React components. Now you should no longer see warnings like `[HMR] Error: Aborted because ... is not accepted`.

Caveats:

* Although suggested by the `react-hot-loader` readme, I did not use the hook support loader as ours seem to work without the addition. You can demonstrate that by adding some simple hook like `useEffect(() => console.log('yo'), ['hot']);` to a component. It should log to console without a hard reload.

## Related Issue





Closes #1590.

## Acceptance 



### Verification Stakeholders
@brendanfalkowski @zetlen 

### Specification



### Verification Steps


After you've got the PR checked out, make some changes (independently) to the following to see the reloading work:
* `createAccountPage.js` - Add a `Button`. Represents a change to a visual component.
* `useCreateAccountPage.js` - Add an effect. Represents a change within a talon/hook.
* `reducers/app.js` - Set `drawer: 'cart'`. Represents a change within the reducer/store.

For each of the above you should see the app perform a swap rather than a full refresh.

## Screenshots / Screen Captures (if appropriate)
Shown above.

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
